c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -0,0 +1,392 @@
+        };
+        Rewrite::Sub(idx, sub_ex.span)
+    }
+}
+
+impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirRewriteVisitor<'a, 'tcx> {
+    type NestedFilter = nested_filter::OnlyBodies;
+
+    fn nested_visit_map(&mut self) -> Self::Map {
+        self.tcx.hir()
+    }
+
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        let mut hir_rw = Rewrite::Identity;
+
+        if let Some(loc) = self.find_primary_location(ex) {
+            self.locations_visited.insert(loc);
+            let rws = self.rewrites.get(&loc).map_or(&[] as &[_], |v| v);
+            for rw in rws {
+                match rw.kind {
+                    mir_op::RewriteKind::OffsetSlice { mutbl } => {
+                        assert!(matches!(hir_rw, Rewrite::Identity));
+                        //assert_eq!(num_args, 2);
+                        // `p.offset(i)` -> `&p[i as usize ..]`
+                        let arr = self.get_subexpr(ex, 0);
+                        let idx = Rewrite::CastUsize(Box::new(self.get_subexpr(ex, 1)));
+                        let elem = Rewrite::SliceTail(Box::new(arr), Box::new(idx));
+                        hir_rw = Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl));
+                    }
+
+                    mir_op::RewriteKind::SliceFirst { mutbl } => {
+                        // `p` -> `&p[0]`
+                        let arr = hir_rw;
+                        let idx = Rewrite::LitZero;
+                        let elem = Rewrite::Index(Box::new(arr), Box::new(idx));
+                        hir_rw = Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl));
+                    }
+
+                    mir_op::RewriteKind::MutToImm => {
+                        // `p` -> `&*p`
+                        let place = Rewrite::Deref(Box::new(hir_rw));
+                        hir_rw = Rewrite::Ref(Box::new(place), hir::Mutability::Not);
+                    }
+
+                    mir_op::RewriteKind::RemoveAsPtr => {
+                        assert!(matches!(hir_rw, Rewrite::Identity));
+                        //assert_eq!(num_args, 1);
+                        hir_rw = self.get_subexpr(ex, 0);
+                    }
+                }
+            }
+        }
+
+        // Emit rewrites on subexpressions first.
+        let applied_mir_rewrite = !matches!(hir_rw, Rewrite::Identity);
+        self.with_materialize_adjustments(applied_mir_rewrite, |this| {
+            intravisit::walk_expr(this, ex);
+        });
+
+        // Materialize adjustments if requested by an ancestor.
+        if self.materialize_adjustments {
+            let adjusts = self.typeck_results.expr_adjustments(ex);
+            assert!(
+                adjusts.is_empty() || !applied_mir_rewrite,
+                "combining adjustments ({:?}) with rewrite ({:?}) is NYI",
+                adjusts,
+                hir_rw,
+            );
+            for adj in adjusts {
+                match adj.kind {
```

Comment:
i think this block should be refactored into its own function

Reply:
Then I'd do a `.fold` over `adjusts` with `hir_rw` as the accumulator.  For the prior `RewriteKind` `match`, too.

Reply:
I factored out the block into a function, but didn't switch to `fold` since I don't think it makes the code any clearer.